Script to run v2 migration commands (#791)
Zipstack/unstract: backend/migrating/v2/README.md

# V2 Migration Guide

This folder contains scripts and management commands for performing data migrations from v1 to v2 for the Unstract Multitenancy system.

> From version `v0.90.3` onwards, there have been signifcant changes in how data is stored. This warrants a data migration and the below steps can come in handy.

Reviewer: The version mentioned here (v0.90.3) is likely to change, as it will reflect the release of v2 without the feature flag. From that point onward, all v2 changes will be fully applied.
cc: @hari-kuriakose

Author: Yes @muhammad-ali-e we need to change it when we introduce that PR - I can leave a placeholder maybe instead?

Reviewer: ok @chandrasekharan-zipstack , We can change it later then.

## Preparing for the migration

### Backup existing data

- Make sure `pg_dump` is installed. This would usually be bundled with postgres and can be installed with

```shell
sudo apt-get install postgresql-client
```

- Take a backup of the existing data with the below command (provide the appropriate options)

```shell
pg_dump -h localhost -p 5432 -U unstract_dev -d unstract_db -f unstract-backup-$(date +"%Y-%m-%d_%H-%M-%S").sql
```

### Create a new database

- Create a new database, for example `unstract_db_v2`

### Dump existing data to this new database

- Exec into the DB if required (from [/docker](/docker/))

```shell
docker exec -it unstract-db /bin/bash
```

- Execute the below command to restore the backup in the new DB

```shell
psql -h localhost -p 5432 -U unstract_dev -d unstract_db_v2 -f unstract-backup-<DateTime>.sql
```

> ### Using a feature flag

Reviewer: We don’t need to mention the feature flag here, as it will be removed in the next few releases. Including it now might cause confusion.

- Check if this migration is used in combination with a feature flag. You can check the code for occurrences of any relevant [FeatureFlag](https://github.com/Zipstack/unstract/blob/0b9981514654634f8b198cdb416b1c9bcd6d423e/backend/backend/constants.py#L33) to confirm

> - If that is the case, head to Flipt configured at `http://localhost:8082` and add a new flag with the same name - in this example of `v2` it is `multi_tenancy_v2`

Reviewer 2: @chandrasekharan-zipstack This is not necessary. Feature flag will be present during development only. Release to end users on being GA will never have any feature flag.
We can remove this section.

## Performing the migration

> The script `migrate-to-v2.sh` in this directory can help perform the migration

Reviewer 2: @chandrasekharan-zipstack 
```suggestion
Run the `migrate-to-v2.sh` script to perform both schema and data migration in oneshot.
```

```shell
./migrate-to-v2.sh
```

Follow the below steps to run each command necessary for the migration explicitly

### Configuring the env

- Make sure to update all the necessary `.env` files

> - For example, in case of `v2` ensure `backend`, `prompt-service`, and `platform-service` `.env` contain `DB_SCHEMA=unstract_v2`

Reviewer 2: @chandrasekharan-zipstack Move this under `## Preparing for the migration` as the second last step.

Author: @hari-kuriakose I've restructured this `README.md` to pose the easier option first with relevant warnings.
Configuring env and creating a new schema are moved to a new section which lists the steps to perform the migration explicitly

### Create a new schema

- Create a new schema within the above DB, for example `unstract_v2`
- This can be done directly in the DB or
- Run the command supported from [/backend/manage.py](/backend/manage.py), for example in case of `v2`

```shell
python manage.py create_v2_schema

> ```

Reviewer 2: @chandrasekharan-zipstack Move this under `## Preparing for the migration` as the last step.

Author: @hari-kuriakose do you intend to place this after the `### Run schema migrations` step? We'll have to create the schema in order to run it and hence kept this and the `.env` configuration before the actualy migration

Author: Refer [this comment](https://github.com/Zipstack/unstract/pull/791#discussion_r1808448134)

### Run schema migrations

- Schema migrations are run by django, this gets invoked when running the backend service
- Or run it explicitly with

```shell
python manage.py migrate
```

- This ensures that the required tables are created or altered

### Run data migrations

- This helps move the existing data from your current schema to the new schema
- It can either be run locally or within the backend container

```shell
docker exec -it unstract-backend /bin/bash
```

- Make sure to activate the virtual environment

```shell
source .venv/bin/activate
```

- Perform the data migration

```shell
python manage.py migrate_to_v2
```

- Test to see if the applied migrations work as expected.

> ## Commands Reference

Reviewer 2: @chandrasekharan-zipstack 
```suggestion
## Django Commands Reference
```

## Performing the migration

> [!TIP]
> If the existing data is important to you, consider [backing up and restoring to a new database](/backend/migrating/v2/README.md#preparing-for-the-migration-safest-approach) beforehand.

Run the `migrate-to-v2.sh` script to perform schema and data migrations in one shot.

```shell
./migrate-to-v2.sh
```

If you prefer explicitly executing the migration commands, refer [these steps below](/backend/migrating/v2/README.md#manually-performing-V2-migration)

## Preparing for the migration (safest approach)

Follow below steps to backup existing data in case of any issues during the migration

### Backup existing data

- Make sure `pg_dump` is installed. This would usually be bundled with postgres and can be installed with

```shell
sudo apt-get install postgresql-client
```

- Take a backup of the existing data with the below command (provide the appropriate options)

```shell
pg_dump -h localhost -p 5432 -U unstract_dev -d unstract_db -f unstract-backup-$(date +"%Y-%m-%d_%H-%M-%S").sql
```

### Create a new database

- Create a new database, for example `unstract_db_v2`

> ### Dump existing data to this new database

Reviewer 2: @chandrasekharan-zipstack 
```suggestion
### Restore backup to the new database
```

## Notes

> - Ensure you have backups of your data before running migration commands.

Reviewer 2: @chandrasekharan-zipstack 
```suggestion
- Ensure you have backed up your data before attempting any migration.
```

- Verify your database configuration and environment variables are correctly set.

> - Review the migration logs for any issues or errors during the migration process.

Reviewer 2: @chandrasekharan-zipstack We could keep this line, it's helpful.
